pkg/controller/replication/replication_controller.go

```diff
@@ -632,5 +634,66 @@ func (rm *ReplicationManager) syncReplicationController(key string) error {
 		updatedRC.Status.AvailableReplicas != *(updatedRC.Spec.Replicas) {
 		rm.enqueueControllerAfter(updatedRC, time.Duration(updatedRC.Spec.MinReadySeconds)*time.Second)
 	}
+
+	pods, err := rm.podLister.Pods(rc.Namespace).List(labels.Set(rc.Spec.Selector).AsSelectorPreValidated())
+	if err != nil {
+		return err
+	}
+	// Find rejected pods and delete them.
+	for _, pod := range pods {
+		if controller.IsPodRejected(pod) {
+			// Ignore deletion error here.
+			rm.podControl.DeletePod(pod.Namespace, pod.Name, rc)
+		}
+	}
+
+	boolPtr := func(b bool) *bool { return &b }
+	controllerRef := &metav1.OwnerReference{
+		APIVersion:         controllerKind.GroupVersion().String(),
+		Kind:               controllerKind.Kind,
+		Name:               rc.Name,
+		UID:                rc.UID,
+		BlockOwnerDeletion: boolPtr(true),
+		Controller:         boolPtr(true),
+	}
+	if pod, err := controller.GetPodFromTemplate(rc.Spec.Template, rc, controllerRef); err == nil {
+		if err = rm.validateNodeConflict(pod); err != nil {
+			return err
+		}
+	}
+
 	return manageReplicasErr
 }
+
+func (rm *ReplicationManager) validateNodeConflict(pod *v1.Pod) error {
+	if len(pod.Spec.NodeName) == 0 {
+		return nil
+	}
+
+	node, err := rm.kubeClient.Core().Nodes().Get(pod.Spec.NodeName, metav1.GetOptions{})
```

Comment:
> Oops... Do we have any conventions for such case? Thanks David!

Generally speaking, we want to separate concerns and make sure that different layers can function independently.  You can think of workload controllers as a bot which keeps the end-user from having to constantly perform the same actions: "I need 5 pods, how many do I have, let me delete/create one".  Thinking about it like that helps reinforce the division of responsibility since end-users aren't generally aware of node or the topology that their kube cluster is running in (see the admin/edit/view RBAC roles).

To help safeguard the division of knowledge further, the e2e tests ought to fail on this because the API server would reject the requests (no RBAC rules).  When you went to add the rule https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go you'd face questions about "least permission required" and "a clever user can use this to do &lt;unexpected thing>" from sig-auth (likely myself or @liggitt).  In this case, it appears that we want the kubelet or scheduler to provide a structured condition on our pod to keep the concerns separated instead of trying to reverse engineer how a particular scheduler may choose to operate.